src/backend/cuda/kernel/memcopy.hpp

```diff
@@ -17,86 +17,248 @@
+        // 3x looping is faster than 4x, 2x remains faster than no looping
+        for (const dim_t i : {3, 4, 5, 7, 11, 2}) {
+            if (elements >= 8192 * i && (mdims[mndims - 1] % i) == 0) {
+                mdims[mndims - 1] /= i;
+                mdims[AF_MAX_DIMS - 1] = i;
+                const dim_t mstride = mdims[mndims - 1] * mstrides[mndims - 1];
+                const dim_t ostride = mdims[mndims - 1] * ostrides[mndims - 1];
+                for (dim_t c = mndims; c < AF_MAX_DIMS; ++c) {
+                    mstrides[c] = mstride;
+                    ostrides[c] = ostride;
+                }
+                mndims = AF_MAX_DIMS;
+                break;  // Perform this operation only once.
+            }
+        }
+    }
+}
 
-    if (ndims == 1) {
-        threads.x *= threads.y;
-        threads.y = 1;
+template<typename T>
+void memcopy(Param<T> out, CParam<T> in, dim_t indims) {
+    bool isLinear  = true;
+    dim_t elements = (indims == 0) ? 0 : 1;
+    for (dim_t dim = 0; dim < indims; ++dim) {
+        isLinear &=
+            (elements == in.strides[dim]) & (elements == out.strides[dim]);
+        elements *= in.dims[dim];
     }
+    if (elements > 0) {
+        const auto stream = cuda::getActiveStream();
+        if (isLinear) {
+            CUDA_CHECK(cudaMemcpyAsync(out.ptr, in.ptr,
+                                       in.elements() * sizeof(T),
+                                       cudaMemcpyDeviceToDevice, stream));
+        } else {
+            const int *maxGridSize =
+                cuda::getDeviceProp(cuda::getActiveDeviceId()).maxGridSize;
+            serializeArray(in.dims, in.strides, indims, out.strides);
+            increaseWorkload(elements, in.dims, in.strides, indims,
+                             out.strides);
+            const dim3 threads = bestBlockSize<dim3>(in.dims, 32);
+            dim3 blocks(divup(static_cast<unsigned>(in.dims[0]), threads.x),
+                        divup(static_cast<unsigned>(in.dims[1]), threads.y),
+                        divup(static_cast<unsigned>(in.dims[2]), threads.z));
+            const bool loop1 = blocks.y > static_cast<unsigned>(maxGridSize[1]);
+            if (loop1) blocks.y = maxGridSize[1];
+            const bool loop2 = blocks.z > static_cast<unsigned>(maxGridSize[2]);
+            if (loop2) blocks.z = maxGridSize[2];
+            EnqueueArgs qArgs(blocks, threads, stream);
 
-    // FIXME: DO more work per block
-    uint blocks_x = divup(in.dims[0], threads.x);
-    uint blocks_y = divup(in.dims[1], threads.y);
+            auto memCopy = common::getKernel("cuda::memcopy", {memcopy_cuh_src},
+                                             {
+                                                 TemplateTypename<T>(),
+                                                 TemplateArg(loop1),
+                                                 TemplateArg(loop2),
+                                             });
```

Comment:
I especially wanted a different kernel for the combinations: <T, NO LOOPS>, <T, LOOP1>, <T,LOOP2>, <T, LOOP1 & LOOP2>.
By passing the LOOP values by value, will I not end up with only combinations <T, LOOP1 & LOOP2>.
This would mean that all cases will have the full overhead, even when not needed.
This will impact the small array's.
In reality, I expect that not all combinations will be used since they are dependent on the slicing of the application.  In most cases only <T, NO LOOPS> will be compiled.

In case I'm wrong, please explain.